File: src/valid-parentheses/valid-parentheses.py
# https://leetcode.com/problems/valid-parentheses

class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        s_open = set(['(', '[', '{'])
        s_close = set([')', ']', '}'])
        stack = list()
        for char in s:
            if char in s_open:
                stack.append(char)
            elif char in s_close:
                if len(stack) == 0:
                    return False
                op = stack.pop()
                if (
                    (char == ']' and op != '[')
                    or (char == ')' and op != '(')
                    or (char == '}' and op != '{')
                ):
                    return False
        if len(stack) > 0:
            return False
        return True

if __name__ == "__main__":
    sol = Solution()
    tests = [
        ("()", True),
        ("()[]{}", True),
        ("(]", False),
    ]
    for (s, solution) in tests:
        result = sol.isValid(s)
        assert solution == result, \
            f"result {result} != solution {solution}, test={s}"
    print("✅ All tests passed")

File: n150/valid-parentheses/valid-parentheses.rs
// https://leetcode.com/problems/valid-parentheses
struct Solution;

use std::collections::VecDeque;

#[derive(Debug)]
enum Position {
    Open(Brace),
    Close(Brace),
}

#[derive(Debug, PartialEq)]
enum Brace {
    Parenthesis,
    Square,
    Angle,
}

impl From<char> for Position {
    fn from(ch: char) -> Self {
        match ch {
            '(' => Self::Open(Brace::Parenthesis),
            ')' => Self::Close(Brace::Parenthesis),
            '[' => Self::Open(Brace::Square),
            ']' => Self::Close(Brace::Square),
            '{' => Self::Open(Brace::Angle),
            '}' => Self::Close(Brace::Angle),
            _ => panic!("invalid bracket character"),
        }
    }
}

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut braces: VecDeque<Position> = s.chars()
            .into_iter()
            .map(|ch| Position::from(ch))
            .collect();

        let mut stack: Vec<Brace> = Vec::with_capacity(braces.len());

        while let Some(pos) = braces.pop_front() {
            match pos {
                Position::Open(b) => {
                    stack.push(b);
                },
                Position::Close(close) => {
                    let open = match stack.pop() {
                        None => return false,
                        Some(open) => open,
                    };
                    if open != close {
                        return false
                    }
                },
            }
        }

        stack.is_empty()
    }
}

fn main() {
    let tests = vec![
        (
            "()[]{}".to_string(),
            true,
        ),
        (
            "(]".to_string(),
            false,
        ),
        (
            "([])".to_string(),
            true,
        ),
        (
            "([)]".to_string(),
            false,
        ),
    ];

    for (s, solution) in tests {
        let result = Solution::is_valid(s);
        assert_eq!(result, solution);
    }

}
